target/scripts/startup/setup.d/security/rspamd.sh

```diff
@@ -294,3 +303,33 @@ function __rspamd__setup_check_authenticated() {
       "${MODULE_FILE}"
   fi
 }
+
+# This function performs a simple check: go through DKIM configuration files, acquire all
+# private key file locations and check whether they exist and whether they can be
+# accesses by Rspamd.
+function __rspamd__check_dkim_permissions() {
+  local DKIM_CONF_FILES DKIM_KEY_FILES
+  [[ -f ${RSPAMD_LOCAL_D}/dkim_signing.conf ]] && DKIM_CONF_FILES+=("${RSPAMD_LOCAL_D}/dkim_signing.conf")
+  [[ -f ${RSPAMD_OVERRIDE_D}/dkim_signing.conf ]] && DKIM_CONF_FILES+=("${RSPAMD_OVERRIDE_D}/dkim_signing.conf")
+
+  # Here, we populate DKIM_KEY_FILES which we later iterate over. DKIM_KEY_FILES
+  # contains all keys files configured by the user.
+  local FILE
+  for FILE in "${DKIM_CONF_FILES[@]}"; do
+    readarray -t DKIM_KEY_FILES_TMP < <(grep -o -E 'path = .*' "${FILE}" | cut -d '=' -f 2 | tr -d ' ";')
+    DKIM_KEY_FILES+=("${DKIM_KEY_FILES_TMP[@]}")
+  done
+
+  for FILE in "${DKIM_KEY_FILES[@]}"; do
+    if [[ -f ${FILE} ]]; then
+      __rspamd__log 'trace' "Checking DKIM file '${FILE}' now"
+      if __do_as_rspamd_user cat "${FILE}" &>/dev/null; then
+        __rspamd__log 'trace' "DKIM file '${FILE}' permissions and ownership appear correct"
```

Comment:
Probably a better way to handle that check than running `cat` like this? @casperklein would probably have good input here.

Presumably using `stat` to check the expected ownership/permissions is what you expect is best (_and more informative to maintainers?_).

If https://github.com/docker-mailserver/docker-mailserver/issues/3630 was implemented (_at least the proposed filesystem layout for keypair files_), you could have the setup scripts copy those over to an internal location and this whole concern would be a non-issue making this check redundant?

`check-for-changes.sh` could additionally monitor the config volume should the keypair change, copying over the update and restarting/reloading the opendkim/rspamd service if needed?

Reply:
> Probably a better way to handle that check than running `cat` like this? @casperklein would probably have good input here.
>
> Presumably using `stat` to check the expected ownership/permissions is what you expect is best (_and more informative to maintainers?_).


This could indeed be more elaborate; but as I said, it ought to be simple for now. We could use `stat` indeed, but then we'd need logic to check all possible cases - remember, we just want the `_rspamd` user to be able to read these files. Matching against permissions is more complicated here. As you also said yourself:

> If #3630 was implemented (_at least the proposed filesystem layout for keypair files_), you could have the setup scripts copy those over to an internal location and this whole concern would be a non-issue making this check redundant?

I want this piece of code to be easy so that we can throw it away once #3630 is implemented. Until then, I actually think this addition is worthwhile.

> `check-for-changes.sh` could additionally monitor the config volume should the keypair change, copying over the update and restarting/reloading the opendkim/rspamd service if needed?

I like this idea! 🚀

Reply:
> We could use `stat` indeed, but then we'd need logic to check all possible cases - remember, we just want the `_rspamd` user to be able to read these files. Matching against permissions is more complicated here.

<details>
<summary>Earlier suggestions</summary>

Check all possible cases?

If you want to ensure the file has specific octal permission (`%a`), or user (`%U`) / group (`%G`):

```bash
if [[ $(stat --format '%a %U %G' "${FILE}" 2>/dev/null) != '775 _rspamd docker' ]]; then
  ...
fi
```

That could work if any of those would be reliable values to match (_easier when we have an internal copy since we can ensure it's copied over with ownership/permissions that we'd expect anyway_).

---

Alternatively, we could do the check that Postfix has with `postfix check` (_as noted in a [recent PR review comment](https://github.com/docker-mailserver/docker-mailserver/pull/3625#discussion_r1385857323)_) for [file permissions with `find`](https://github.com/vdukhovni/postfix/blob/2047994eeee1dbf0d8c0537ba00acb8986a8a370/postfix/conf/postfix-script#L352-L354):

```bash
if [[ $(find "${FILE}" -perm -ugo=r 2>/dev/null) ]]; then
  ...
fi
```

This uses symbolic notation instead of octal, bit easier to grok, but `-444` would be equivalent AFAIK. The `-` prefix for the value ensures that the file has these baseline permission for all user/group/other ownership (_minimum perms, rather than for any ownership or exact permissions_):

```
-perm MASK      At least one mask bit (+MASK), all bits (-MASK),
                        or exactly MASK bits are set in file's mode
```

That would still better communicate the expectation of the file being readable. Although I get that your expectation is laxed, the file could have user or group ownership for `_rspamd` and thus not need to be readable by everybody, but that would also be valid...and yet you can't use `+ugo=r` (_at least one is readable_) since that may not apply to `_rspamd` 🤷‍♂️ 

</details>

---

This should technically be valid:

```bash
# Internally _rspamd would only be valid as user/group,
# otherwise require read permission for `other` perms:
if [[ $(find "${FILE}" -user _rspamd -or -group _rspamd -or -perm -o=r 2>/dev/null) ]]; then
  ...
fi
```

> This could indeed be more elaborate; but as I said, it ought to be simple for now.
> I want this piece of code to be easy so that we can throw it away once https://github.com/docker-mailserver/docker-mailserver/issues/3630 is implemented.

If that's likely to be addressed before the end of the year, that seems fine 👍 

Otherwise the `find` approach seems simple enough and conveys expectation a bit better?

```suggestion
      if [[ $(find "${FILE}" -user _rspamd -or -group _rspamd -or -perm -o=r 2>/dev/null) ]]; then
        __rspamd__log 'trace' "DKIM file '${FILE}' permissions and ownership appear correct"
```

Reply:
This would not quite work as expected, but I like it, so I changed it:

```bash
# See https://serverfault.com/a/829314 for an explanation on `-exec false {} +`
# We additionally resolve symbolic links to check the permissions of the actual files
if find "$(realpath -eL "${FILE}")" -user _rspamd -or -group _rspamd -or -perm -o=r -exec false {} +; then
  __rspamd__log 'warn' "Rspamd DKIM private key file '${FILE}' does not appear to have correct permissions/ownership for Rspamd to use it"
else
  __rspamd__log 'trace' "DKIM file '${FILE}' permissions and ownership appear correct"
fi
```